packages/amplify-graphql-api-construct/src/internal/cdk-shim.ts

```diff
@@ -0,0 +1,145 @@
+): void => {
+  const {
+    schema,
+    resolvers,
+    functions,
+    rootStack,
+    stacks,
+  } = params;
+
+  const resolverDir = fs.mkdtempSync(path.join(assetDir, RESOLVERS_DIRECTORY));
+
+  const schemaPath = path.join(assetDir, SCHEMA_FILE_NAME);
+  fs.writeFileSync(schemaPath, schema);
+  const schemaAsset = new Asset(scope, 'schema-asset', { path: schemaPath });
+  const schemaResourceKey = Object.entries(rootStack.Resources ?? {}).find(([_, resource]) => resource.Type === 'AWS::AppSync::GraphQLSchema')?.[0];
+  const definitionLocation = ['s3:/', schemaAsset.s3BucketName, schemaAsset.s3ObjectKey].join('/');
+  rootStack.Resources![schemaResourceKey!].Properties.DefinitionS3Location = definitionLocation;
+
+  // There's probably a more elegant way to do this, but basically rewrite any sort of paths
+  // (be they asset or otherwise) in order to point them to the asset directory for the CDK.
+  Object.entries(resolvers).forEach(([resolverName, resolverCode]) => {
+    const resolverPath = path.join(resolverDir, resolverName);
+    fs.writeFileSync(resolverPath, resolverCode);
+    const currSchemaAsset = new Asset(scope, `resource-${resolverName}-asset`, { path: resolverPath });
+    const currDefinitionLocation = ['s3:/', currSchemaAsset.s3BucketName, currSchemaAsset.s3ObjectKey].join('/');
+    Object.keys(stacks).forEach((stackName) => {
+      /* eslint-disable max-len */
+      const requestResponseMappingResourceKey = Object.entries(stacks[stackName].Resources ?? {}).find(([_, resource]) => resource?.Properties?.RequestMappingTemplateS3Location && JSON.stringify(resource.Properties.RequestMappingTemplateS3Location).includes(resolverName))?.[0];
+      const responseResponseMappingResourceKey = Object.entries(stacks[stackName].Resources ?? {}).find(([_, resource]) => resource?.Properties?.ResponseMappingTemplateS3Location && JSON.stringify(resource.Properties.ResponseMappingTemplateS3Location).includes(resolverName))?.[0];
+      const codeResourceKey = Object.entries(stacks[stackName].Resources ?? {}).find(([_, resource]) => resource?.Properties?.CodeS3Location && JSON.stringify(resource.Properties.CodeS3Location).includes(resolverName))?.[0];
+      /* eslint-enable max-len */
+
+      if (requestResponseMappingResourceKey) {
+        stacks[stackName].Resources![requestResponseMappingResourceKey!].Properties
+          .RequestMappingTemplateS3Location = currDefinitionLocation;
+      }
+      if (responseResponseMappingResourceKey) {
+        stacks[stackName].Resources![responseResponseMappingResourceKey!].Properties
+          .ResponseMappingTemplateS3Location = currDefinitionLocation;
+      }
+      if (codeResourceKey) {
+        stacks[stackName].Resources![codeResourceKey!].Properties
+          .CodeS3Location = currDefinitionLocation;
+      }
+    });
+  });
+
+  // Look for AWS::Lambda::Function, and rewrite the path.
+  // This is just terrible.
```

Comment:
Is there any chance that this while `cdk-shim` could be eliminated by cutting functionality scope of the construct? And bring it back when implementation is refactored to not need this.

I'm a bit skeptical about all this disk access , rewrites and lookups outside of normal sythesis process (if I understand this correctly...).

Do these tricks here impact construct usability in plain CDK app or stuff like cdk watch or cdk change set computation ?

Reply:
Without this we basically can't ship anything, since all our core resources (graphql schema + resolvers + functions) are dependent on file assets.

I've done a fair bit of internal testing using cdk synth and watch and haven't experienced any particular issues w/ this approach so far.